source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -0,0 +1,184 @@
+  CONSTRUCT_ON_FIRST_USE(int, []() -> int {
+    int ssl_index = SSL_get_ex_new_index(0, nullptr, nullptr, nullptr, nullptr);
+    RELEASE_ASSERT(ssl_index >= 0);
+    return ssl_index;
+  }());
+}
+
+} // namespace
+
+Config::Config(Stats::Scope& scope)
+    : stats_{ALL_TLS_INSPECTOR_STATS(POOL_COUNTER_PREFIX(scope, "tls_inspector."))},
+      ssl_ctx_(SSL_CTX_new(TLS_with_buffers_method())) {
+  SSL_CTX_set_options(ssl_ctx_.get(), SSL_OP_NO_TICKET);
+  SSL_CTX_set_session_cache_mode(ssl_ctx_.get(), SSL_SESS_CACHE_OFF);
+  SSL_CTX_set_tlsext_servername_callback(
+      ssl_ctx_.get(), [](SSL* ssl, int* out_alert, void*) -> int {
+        Filter* filter = static_cast<Filter*>(SSL_get_ex_data(ssl, sslIndex()));
+        filter->onServername(SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name));
+
+        // Return an error to stop the handshake; we have what we wanted already.
+        *out_alert = SSL_AD_USER_CANCELLED;
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+      });
+}
+
+bssl::UniquePtr<SSL> Config::newSsl() { return bssl::UniquePtr<SSL>{SSL_new(ssl_ctx_.get())}; }
+
+Filter::Filter(const ConfigSharedPtr config) : Filter(config, TLS_DEFAULT_MAX_CLIENT_HELLO) {}
+
+Filter::Filter(const ConfigSharedPtr config, uint32_t max_client_hello_size)
+    : config_(config), ssl_(config_->newSsl()) {
+  SSL_set_ex_data(ssl_.get(), sslIndex(), this);
+  SSL_set_accept_state(ssl_.get());
+
+  // TODO(ggreenway) PERF: Put a buffer in thread-local storage and have every Filter share
+  // the buffer.
+  buf_.reserve(max_client_hello_size);
+}
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "tls inspector: new connection accepted");
+  Network::ConnectionSocket& socket = cb.socket();
+  ASSERT(file_event_.get() == nullptr);
+
+  // TODO(ggreenway): detect closed connections.
+  file_event_ =
+      cb.dispatcher().createFileEvent(socket.fd(),
+                                      [this](uint32_t events) {
+                                        ASSERT(events == Event::FileReadyType::Read);
+                                        onRead();
+                                      },
+                                      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+
+  // TODO(PiotrSikora): make this configurable.
+  timer_ = cb.dispatcher().createTimer([this]() -> void { onTimeout(); });
+  timer_->enableTimer(std::chrono::milliseconds(15000));
+
+  cb_ = &cb;
+  return Network::FilterStatus::StopIteration;
+}
+
+void Filter::onServername(absl::string_view name) {
+  if (!name.empty()) {
+    config_->stats().sni_found_.inc();
+    cb_->socket().setRequestedServerName(name);
+  } else {
+    config_->stats().no_sni_found_.inc();
+  }
+  servername_set_ = true;
+}
+
+void Filter::onRead() {
+  Network::ConnectionSocket& socket = cb_->socket();
+
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  ssize_t n = os_syscalls.recv(socket.fd(), buf_.data(), buf_.capacity(), MSG_PEEK);
```

Comment:
My hope was that if more data came in, we'd get another event, but we certainly need to double-check that. If that's not the case, I don't know how this can be made to work without polling (which would be awful).

Reply:
Yeah, I've been walking around my office trying to find someone familiar enough with the kernel internals to give an authoritative answer here.   If the kernel is doing epoll edge triggered correctly (i.e. only notifying you of an edge) you'd only get an event on first data and because you never drained the data you'd not get a subsequent event.  I have at least one coworker convinced the actual implementation gives more enthusiastic notifications than that but that's opinion based not code based.  Worst case we may have to dig through kernel source :-/

What I can say authoritatively is I'm fairly sure I have experience with peeking-without-polling in production for O(hundred bytes) without problems, so if SNI is likely to be in that range and it turns out the kernel does true edge triggering we could possibly comment this up as an implementation flaw ("does not handle client hellos fragmented across multiple packets") and call it good enough for now

Reply:
I can look at the kernel code if need be to answer for sure, but I'm fairly positive that in edge triggered scenarios, you need to read until you receive EAGAIN to get another notification. Whether you actually read the data or use MSG_PEEK I don't think matters. This is what we already do here: https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L126

As long as subsequence filters (or the connection) always attempt to read until EAGAIN when they run I think we are OK.

Reply:
I agree.  Having double checked I think my internal experience relies on packets not being fragmented past 4 bytes which is fairly reasonable (if you can't get more than 3 bytes to a packet you have worst problems).  Expecting full SNI is less safe.  I suspect we either need to peek-and-pull or read and hand off already-read data to the final consumer.  Messy either way.

Reply:
This is probably OK w.r.t peek + EAGAIN (though a unit test would be great to confirm)

There's one other related issue, is that you're still not necessarily reading until EAGAIN.   in the case of TLS/1.3 0-rtt handshake, the initial payload may be larger than TLS_DEFAULT_MAX_CLIENT_HELLO    My remaining concern is that if we consume the EAGAIN, the next layer to pick up the connection may not get the notification and so may not read.    Again @mattklein123 may know offhand but I'm not sure if we explicitly read on new incoming connections or if we rely on the initial EPOLLIN (which we might not be preserving in this case)

Reply:
Just to summarize for everyone a lengthy offline Slack conversation, here is what I think is happening, some confusion, and some areas for further investigation.

1) I'm fairly certain that if you read until EAGAIN, whether peek or not, you will get another event. I don't see how it could work otherwise. But would love to confirm.
2) The larger confusion is I realized that we *don't* attempt a read when we hand off the fd to the real connection handling code. What we do instead is re-register with epoll. Now, given that in the case of proxy proto it will almost always be the case that there is more data then we needed and we didn't read until EAGAIN, I was confused about how this ever worked. I suspect that when you re-register with epoll using ET, it will act as LT for the first registration,  and raise an initial event.

I would love to confirm (1) and (2) above with some UTs if possible...

I think the TLDR is that we need to use the following rules:
1) Each filter or ultimate connection can't assume that they will receive another event.
2) They either have to attempt to read, or re-register with epoll which will trigger an event. (?)

Reply:
Note that there isn't more data coming after ClientHello from the client, since it's expecting ServerHello response from the server, which we don't do at this point. This means that there won't be any future events delivered by the kernel.

However, this works, so it seems that registering for events again in `ConnectionImpl::ConnectionImpl()` takes care of that.

Reply:
SGTM.  I think my take-away is this is confusing and worth documenting for posterity (@ggreenway can you try to find an appropriate place?).  (optional) bonus points for integration tests (really I'm always a fan of integration tests :-P)